src/vmm/src/guest_config/x86_64/mod.rs

```diff
@@ -1,17 +1,378 @@
+    #[error("Can create cpuid from raw: {0}")]
+    CpuidFromRaw(super::cpuid::CpuidTryFromRawCpuid),
+}
 
 /// CPU configuration for x86_64 CPUs
-#[derive(Clone, Debug, Eq, PartialEq)]
+#[derive(Debug, Clone, PartialEq)]
 pub struct X86_64CpuConfiguration {
     /// CPUID configuration
     pub cpuid: Cpuid,
     /// Register values as a key pair for model specific registers
     /// Key: MSR address
     /// Value: MSR value
     pub msrs: HashMap<u32, u64>,
+
+    /// Set of supported MSRs
+    pub msr_list: HashSet<u32>,
+
+    /// Architectural MSPs required for boot
+    pub msr_boot_entries: Vec<kvm_msr_entry>,
+}
+
+impl X86_64CpuConfiguration {
+    /// Craetes new CpuConfig with default values
+    pub fn new(vm: &Vm) -> Result<Self, Error> {
+        let supported_cpuid = vm.supported_cpuid().clone();
+        let supported_cpuid =
+            Cpuid::try_from(RawCpuid::from(supported_cpuid)).map_err(Error::CpuidFromRaw)?;
+        let msr_list = HashSet::default();
+
+        let msr_boot_entries = create_boot_msr_entries();
+
+        Ok(Self {
+            cpuid: supported_cpuid,
+            msrs: Default::default(),
+            msr_list,
+            msr_boot_entries,
+        })
+    }
+
+    /// Craetes new CpuConfig with cpu template changes applied
+    pub fn from_cpu_template(vm: &Vm, template: &CpuTemplateType) -> Result<Self, Error> {
+        let supported_cpuid = vm.supported_cpuid().clone();
+        let mut supported_cpuid =
+            Cpuid::try_from(RawCpuid::from(supported_cpuid)).map_err(Error::CpuidFromRaw)?;
+        let mut msr_list = vm.supported_msrs().as_slice().iter().copied().collect();
+
+        // Initialize some architectural MSRs that will be set for boot.
+        let mut msr_boot_entries = create_boot_msr_entries();
+
+        match template {
+            CpuTemplateType::Custom(template) => {
+                // TODO QUERY ADDITIONAL MSRS
+                let config = Self {
+                    cpuid: supported_cpuid,
+                    msrs: Default::default(),
+                    msr_list,
+                    msr_boot_entries,
+                };
+
+                Self::with_template(config, template)
+            }
+            CpuTemplateType::Static(template) => {
+                // If a template is specified, get the CPUID template, else use `cpuid`.
+                let config_cpuid = match template {
+                    CpuFeaturesTemplate::C3 => cpuid_templates::c3::template(),
+                    CpuFeaturesTemplate::T2 => cpuid_templates::t2::template(),
+                    CpuFeaturesTemplate::T2S => cpuid_templates::t2s::template(),
+                    CpuFeaturesTemplate::T2CL => cpuid_templates::t2cl::template(),
+                    CpuFeaturesTemplate::T2A => cpuid_templates::t2a::template(),
+                    CpuFeaturesTemplate::None | CpuFeaturesTemplate::Custom => {
+                        unreachable!("None and Custom states are invalid")
```

Comment:
Invalidating `None` changes existing behaviour that isn't related/required for the core change of this PR. The `GET /machine-config` API will expect a value if a template isn't specified by the user.

Reply:
In this specific part of code `CpuFeaturesTemplate::None` is invalid, because this code never executes if `CpuFeaturesTemplate::None` is set.